File: src/services/tts.ts
import { google, texttospeech_v1 } from 'googleapis';
import { optionalEnv } from '@/lib/env';
import { getGoogleCredentials } from '@/lib/google-credentials';
import { Language } from './sheets';

/** Google Cloud TTS locale and voice per language. Override with `TTS_VOICE_<LANG>`. */
const VOICES: Record<Language, { languageCode: string; name: string }> = {
  ja: { languageCode: 'ja-JP', name: 'ja-JP-Chirp3-HD-Enceladus' },
  en: { languageCode: 'en-US', name: 'en-US-Neural2-F' },
  es: { languageCode: 'es-ES', name: 'es-ES-Neural2-A' },
  pt: { languageCode: 'pt-BR', name: 'pt-BR-Neural2-A' },
  id: { languageCode: 'id-ID', name: 'id-ID-Standard-A' },
  ar: { languageCode: 'ar-XA', name: 'ar-XA-Wavenet-A' },
};

export class TextToSpeechService {
  private tts: texttospeech_v1.Texttospeech;

  constructor() {
    const auth = new google.auth.GoogleAuth({
      credentials: getGoogleCredentials(),
      scopes: ['https://www.googleapis.com/auth/cloud-platform'],
    });
    this.tts = google.texttospeech({ version: 'v1', auth });
  }

  /**
   * Synthesizes narration as MP3. `speakingRate` is how the render pipeline
   * fits the narration into the 20s / 65s slot.
   */
  async synthesize(text: string, language: Language, speakingRate = 1): Promise<Buffer> {
    const fallback = VOICES[language];
    const voice = {
      languageCode: optionalEnv(`TTS_LANGUAGE_CODE_${language.toUpperCase()}`) ?? fallback.languageCode,
      name: optionalEnv(`TTS_VOICE_${language.toUpperCase()}`) ?? fallback.name,
    };

    const response = await this.tts.text.synthesize({
      requestBody: {
        input: { text },
        voice,
        audioConfig: {
          audioEncoding: 'MP3',
          speakingRate: Math.min(Math.max(speakingRate, 0.5), 2),
        },
      },
    });

    const audio = response.data.audioContent;
    if (!audio) throw new Error('Text-to-Speech returned no audio');
    return Buffer.from(audio, 'base64');
  }
}

File: renderer/tts.py
"""Per-sentence narration, so each line appears exactly when it is spoken."""

import os

from google.cloud import texttospeech

from text import apply_reading_hints, split_for_speech

VOICES: dict[str, tuple[str, str]] = {
    "ja": ("ja-JP", "ja-JP-Chirp3-HD-Enceladus"),
    "en": ("en-US", "en-US-Chirp3-HD-Enceladus"),
    "es": ("es-ES", "es-ES-Neural2-A"),
    "pt": ("pt-BR", "pt-BR-Neural2-A"),
    "id": ("id-ID", "id-ID-Standard-A"),
    "ar": ("ar-XA", "ar-XA-Wavenet-A"),
    "fr": ("fr-FR", "fr-FR-Chirp3-HD-Enceladus"),
    "de": ("de-DE", "de-DE-Chirp3-HD-Enceladus"),
}

# The speech API rejects a sentence it considers too long; Japanese reaches that first.
SENTENCE_LIMIT: dict[str, int] = {"ja": 80, "default": 280}


class Narrator:
    def __init__(self) -> None:
        self._client = texttospeech.TextToSpeechClient()

    def synthesize(self, text: str, language: str, path: str, speaking_rate: float = 1.0) -> str:
        """Speed is set on the voice itself; ffmpeg's `atempo` mangles the timestamps mixing needs."""
        language_code, name = VOICES[language]
        language_code = os.environ.get(f"TTS_LANGUAGE_CODE_{language.upper()}", language_code)
        name = os.environ.get(f"TTS_VOICE_{language.upper()}", name)

        voice = texttospeech.VoiceSelectionParams(language_code=language_code, name=name)
        config = texttospeech.AudioConfig(
            audio_encoding=texttospeech.AudioEncoding.MP3, speaking_rate=speaking_rate
        )
        limit = SENTENCE_LIMIT.get(language, SENTENCE_LIMIT["default"])
        audio = b""
        for piece in split_for_speech(apply_reading_hints(text, language), limit):
            response = self._client.synthesize_speech(
                input=texttospeech.SynthesisInput(text=piece), voice=voice, audio_config=config
            )
            audio += response.audio_content
        with open(path, "wb") as file:
            file.write(audio)
        return path
